src/array_unique.rs

```diff
@@ -0,0 +1,49 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+
+use hashbrown::HashMap;
+use numpy::PyReadonlyArray2;
+
+#[pyfunction]
+#[pyo3(text_signature = "(array)")]
+pub fn unique(array: PyReadonlyArray2<u16>) -> PyResult<(Vec<usize>, Vec<usize>)> {
+    let array = array.as_array();
+    let shape = array.shape();
+    let mut table = HashMap::<Vec<u16>, usize>::new();
+    let mut indexes = vec![0; 0];
+    let mut inverses = vec![0; shape[0]];
+    for i in 0..shape[0] {
+        let mut v = vec![0u16; shape[1]];
+        for j in 0..shape[1] {
+            v[j] = array[(i, j)];
+        }
+        match table.get(&v) {
+            Some(id) => inverses[i] = *id,
+            None => {
+                let new_id = table.len();
+                table.insert(v, new_id);
+                inverses[i] = new_id;
+                indexes.push(i);
+            }
+        }
+    }
+    Ok((indexes, inverses))
```

Comment:
If you return a tuple of `Vec<usize>` here you'll end up copying each array by iterating over them and converting each element from a usize to a python int and making a python list from it. If you're don't need to modify things it's more efficient to convert the `Vec` into a numpy array and return that. This avoid the iteration and copy because we're creating an immutable numpy array that refers to the memory for the vec on the rust heap. Depending on the size of the output vectors this can make a big difference. I did this for the edge collection return here: https://github.com/Qiskit/qiskit-terra/blob/main/src/edge_collections.rs#L50-L58 (you shouldn't need the `clone()` because you are just moving the Vec to the output pyarray, while that method the struct/pyclass still owns the edges member so we need to clone)